File: backend/app/services/platforms.py
"""Platform config service — DB-backed with a warm sync cache.

Dimensions drive every render and every request. ``get_format_info`` and
``validate_platforms`` are synchronous, so this service keeps a small TTL
cache refreshed from the ``platforms`` table: at lifespan, on every API
write, and once per ``generate_task`` (workers never run the FastAPI
lifespan). Before the first refresh the seed YAML is the fallback, so unit
tests and pre-boot reads still resolve dimensions.
"""

from __future__ import annotations

import logging
import time
from dataclasses import dataclass

from app.config import get_settings

log = logging.getLogger(__name__)

_CACHE_TTL = 5.0
_platform_cache: list[dict] | None = None
_platform_cache_ts = 0.0

_VALID_FAMILIES = {"square", "portrait", "story", "landscape"}


@dataclass
class PlatformInfo:
    id: str
    name: str
    width: int
    height: int
    family: str
    is_active: bool = True


def platform_to_dict(row) -> dict:
    from app.core.time import iso_utc

    return {
        "id": row.id,
        "name": row.name,
        "width": row.width,
        "height": row.height,
        "family": row.family,
        "is_active": bool(row.is_active),
        "sort_order": row.sort_order,
        "created_at": iso_utc(row.created_at),
        "updated_at": iso_utc(row.updated_at),
    }


def _yaml_seed() -> list[dict]:
    """Seed-platform fallback (pre-boot / unit tests, before DB is loaded).

    Family mirrors the DB seed: design-instruction ``format_families`` first,
    aspect heuristic otherwise.
    """
    from pathlib import Path

    from app.services.design_instruction import load_design_instruction
    from app.services.tokens import load_platforms

    settings = get_settings()
    dims = load_platforms(settings.platforms_path)
    di = load_design_instruction(Path(settings.design_system_dir) / "design-instruction.yaml")
    families = di.get("format_families", {}) if isinstance(di, dict) else {}
    out = []
    for i, (pid, (w, h)) in enumerate(sorted(dims.items())):
        fam = families.get(pid)
        if fam not in ("square", "portrait", "story", "landscape"):
            fam = "square" if h <= w else "portrait"
        out.append(
            {
                "id": pid,
                "name": pid.replace("-", " ").title(),
                "width": w,
                "height": h,
                "family": fam,
                "is_active": True,
                "sort_order": i,
            }
        )
    return out


async def refresh_platforms(pool=None) -> None:
    """Reload the platform cache from the DB (lifespan, writes, workers)."""
    global _platform_cache, _platform_cache_ts
    try:
        from app.db.repositories.platforms import PlatformRepository
        from app.db.session import get_shared_session_factory

        pool = pool or (await get_shared_session_factory())
        async with pool() as session:
            rows = await PlatformRepository(session).list(include_inactive=True)
        _platform_cache = [platform_to_dict(r) for r in rows]
        _platform_cache_ts = time.monotonic()
    except Exception as e:  # noqa: BLE001
        log.warning("[platforms] refresh failed: %s", e)


def _platforms() -> list[dict]:
    global _platform_cache, _platform_cache_ts
    now = time.monotonic()
    if _platform_cache is not None:
        if now - _platform_cache_ts < _CACHE_TTL:
            return _platform_cache
        return _platform_cache  # stale is fine; async refresh happens on writes
    return _yaml_seed()


def list_platforms(include_inactive: bool = False) -> list[dict]:
    rows = _platforms()
    if include_inactive:
        return rows
    return [r for r in rows if r.get("is_active", True)]


def get_platform(platform_id: str) -> dict | None:
    for r in _platforms():
        if r["id"] == platform_id:
            return r
    return None


def get_platform_dims(platform_id: str) -> tuple[int, int] | None:
    row = get_platform(platform_id)
    if row is None:
        return None
    return int(row["width"]), int(row["height"])


def family_of(platform_id: str) -> str:
    row = get_platform(platform_id)
    if row and row.get("family") in _VALID_FAMILIES:
        return row["family"]
    # Carousel slide ids (instagram-carousel-N / instagram-carousel-portrait-N)
    # resolve to their base platform's family.
    import re as _re

    m = _re.match(r"^(.+)-(\d+)$", platform_id)
    base = m.group(1) if m else platform_id
    row = get_platform(base)
    if row and row.get("family") in _VALID_FAMILIES:
        return row["family"]
    dims = get_platform_dims(base)
    if dims:
        w, h = dims
        return "square" if h <= w else "portrait"
    return "square"

File: frontend/src/lib/platforms.ts
// DB-backed platform config. The Studio owns the rows; this module keeps a
// warm module cache so sync helpers (formatDims / FAMILY_DIMS) work everywhere
// without threading async state through every consumer. `loadPlatforms()`
// refreshes it; `usePlatforms()` also triggers a re-render on change.

import type { PlatformInfo } from "./api"
import { listPlatforms } from "./api"

export type { PlatformInfo }

export interface Dimensions {
  width: number
  height: number
}

const SQUARE: Dimensions = { width: 1080, height: 1080 }

// Fallback defaults — correct for the seed system; overwritten on load.
export const FAMILY_DIMS: Record<string, Dimensions> = {
  square: SQUARE,
  portrait: { width: 1080, height: 1350 },
  story: { width: 1080, height: 1920 },
  landscape: { width: 1200, height: 627 },
}

export const FORMAT_DIMS: Record<string, Dimensions> = {
  "instagram-square": SQUARE,
  "instagram-carousel": SQUARE,
  "instagram-carousel-portrait": { width: 1080, height: 1350 },
  "instagram-portrait": { width: 1080, height: 1350 },
  "instagram-story": { width: 1080, height: 1920 },
  "linkedin-post": { width: 1200, height: 627 },
  "twitter-card": { width: 1200, height: 675 },
  "facebook-post": { width: 1200, height: 630 },
  "pinterest-pin": { width: 1000, height: 1500 },
}

const _SLIDE_RE = /^(instagram-carousel(?:-portrait)?)-(\d+)$/

/** Fetch platforms and populate the module caches. */
export async function loadPlatforms(): Promise<PlatformInfo[]> {
  const rows = await listPlatforms()
  const byId: Record<string, PlatformInfo> = {}
  // Build a family → first-active-platform map once (instead of a find per
  // family in the loop below).
  const firstByFamily: Record<string, PlatformInfo> = {}
  for (const row of rows) {
    byId[row.id] = row
    FORMAT_DIMS[row.id] = { width: row.width, height: row.height }
    if (!(row.family in firstByFamily)) firstByFamily[row.family] = row
  }
  for (const fam of Object.keys(FAMILY_DIMS)) {
    const first = firstByFamily[fam]
    if (first) FAMILY_DIMS[fam] = { width: first.width, height: first.height }
  }
  return rows
}

/** Current platform ids (includes the default seed set before load). */
export function knownPlatforms(): string[] {
  return Object.keys(FORMAT_DIMS)
}

export function platformDims(format: string): Dimensions | null {
  const exact = FORMAT_DIMS[format]
  if (exact) return exact
  const m = _SLIDE_RE.exec(format)
  if (m && FORMAT_DIMS[m[1]]) return FORMAT_DIMS[m[1]]
  return null
}

export function formatDims(format: string): Dimensions {
  return platformDims(format) ?? SQUARE
}

export function familyOfPlatform(format: string): string {
  const m = _SLIDE_RE.exec(format)
  const base = m ? m[1] : format
  const d = platformDims(base)
  if (!d) return "square"
  return d.height > d.width ? "portrait" : d.width > d.height ? "landscape" : "square"
}
